src/libraries/Common/src/System/IO/Archiving.Utils.cs

```diff
@@ -32,26 +31,36 @@ public static string EntryFromPath(string entry, int offset, int length, ref cha
             }
 
             if (length == 0)
+            {
                 return appendPathSeparator ? PathSeparatorString : string.Empty;
+            }
 
-            int resultLength = appendPathSeparator ? length + 1 : length;
-            EnsureCapacity(ref buffer, resultLength);
-            entry.CopyTo(offset, buffer, 0, length);
-
-            // '/' is a more broadly recognized directory separator on all platforms (eg: mac, linux)
-            // We don't use Path.DirectorySeparatorChar or AltDirectorySeparatorChar because this is
-            // explicitly trying to standardize to '/'
-            for (int i = 0; i < length; i++)
+            if (appendPathSeparator)
             {
-                char ch = buffer[i];
-                if (ch == Path.DirectorySeparatorChar || ch == Path.AltDirectorySeparatorChar)
-                    buffer[i] = PathSeparatorChar;
+                length++;
             }
 
-            if (appendPathSeparator)
-                buffer[length] = PathSeparatorChar;
+            return string.Create(length, (appendPathSeparator, offset, entry), static (dest, state) =>
+            {
+                state.entry.AsSpan(state.offset).CopyTo(dest);
+
+                // '/' is a more broadly recognized directory separator on all platforms (eg: mac, linux)
+                // We don't use Path.DirectorySeparatorChar or AltDirectorySeparatorChar because this is
+                // explicitly trying to standardize to '/'
+                for (int i = 0; i < dest.Length; i++)
+                {
+                    char ch = dest[i];
+                    if (ch == Path.DirectorySeparatorChar || ch == Path.AltDirectorySeparatorChar)
```

Comment:
If on Unix both these constants are `/`
https://github.com/dotnet/runtime/blob/e71a9583b4d6c9bd97edd87cda7f98f232f63530/src/libraries/Common/src/System/IO/PathInternal.Unix.cs#L13-L14
we could skip the cycle on Unix. On Windows we could do only one check `\` in the cycle.

Reply:
This code is the same on both platforms.

@EgorBo just out of curiosity, would the JIT in theory be able to legally collapse such a thing? ie, `if (ch == '/' || ch == '/') ` ..

Reply:
> This code is the same on both platforms.

It makes no sense to replace `/` with `/` on Unix.

Reply:
I get that, but what are you proposing -- duplicate this method for Unix and Windows so they can be different? Is this code path that hot?

Reply:
It's not super hot, but it can be improved.  I need to push up a fix anyway, so I'll do so.